AraEvent/AraEventCalibrator.cxx

```diff
@@ -1458,11 +1491,159 @@ Bool_t AraEventCalibrator::TrimFirstBlock(UsefulAtriStationEvent *theEvent, std:
+                
+                const int num_block = int(blockStarts.size());
+                std::vector<double> dT(num_block); // width of block
+                for(int i = 0; i < num_block; ++i) 
+                    dT[i] = blockEnds[i] - blockStarts[i]; 
+
+                // interpolate onto regular 0.5 ns timestep
+                TGraph* gr = new TGraph(total_samples, &tmpT[0], &tmpV[0]); 
+                const double dt = 0.5; // ns
+                TGraph* grInt = FFTtools::getInterpolatedGraph(gr, dt);
+                const int len = grInt->GetN();
+                double* trace = grInt->GetY();
+                
+                // transform to frequency space 
+                FFTWComplex* spec = FFTtools::doFFT(len, trace);                
+ 
+                // get spectrum data to fit and precalculate constants for fit
+                const int nFreq = len/2+1;
+                const double dFreq = 1./dt/len; // GHz
+                const double lowPassFreq = 50.e-3; // GHz 
+
+                // calculate the "data" to fit to
+                std::vector<double> angFreq;
+                std::vector<double> specMag;
+                std::vector<double> specPhase;
+                for(int i = 0; i < nFreq; ++i) {
+                    const double thisFreq = i*dFreq;
+                    if(thisFreq <= lowPassFreq) { // only use low frequency data
+                        angFreq.push_back(2.*M_PI*thisFreq);
+
+                        specMag.push_back(spec[i].getAbs());
+                        specPhase.push_back(spec[i].getPhase());
+                    }
+                }
+                
+                // calculate the "model" whose parameters we solve for
+                const int nAngFreq = int(angFreq.size()); // number of frequencies in low frequency data
+                std::vector< std::vector<double> > constMag(num_block, std::vector<double>(nAngFreq));
+                std::vector< std::vector<double> > constPhase(num_block, std::vector<double>(nAngFreq));
+                for(int b = 0; b < num_block; ++b) {
+                    const double blockWidth = dT[b];
+                    const double blockStartShift = blockStarts[b] - blockStarts[0];
+                    for(int i = 0; i < nAngFreq; ++i) {
+                        const double thisAngFreq = angFreq[i];
+
+                        if(thisAngFreq == 0)
+                            constMag[b][i] = blockWidth/dt;
+                        else
+                            constMag[b][i] = TMath::Sin(blockWidth*thisAngFreq/2.)/TMath::Sin(dt*thisAngFreq/2.); 
+                        constPhase[b][i] = -thisAngFreq*(blockWidth/2.+blockStartShift);
+                    }
+                }
+
+                // get block offsets
+                // we do this by minimizing the squared error between the low passed spectrum and the block offsets
+                // this can be done by solving a matrix equation, so we start by setting that up
```

Comment:
You can solve the complex system by just using a matrix twice as big, right?

Reply:
Agreed. Do you think that's advantageous? Or maybe this is just a comment.